Adding Query Params, Headers, and Form Data with Null Values throws NullPointerException

Example with ```addHeader()```:

```groovy
def foo() {
    HttpRequest httpRequest = new HttpRequest()
    httpRequest.addHeader('foo', null)
}
```

Example with ```addQueryParameter()```:

```groovy
def foo() {
    HttpRequest httpRequest = new HttpRequest()
    httpRequest.addQueryParameter('foo', null)
}
```

Example with ```addField()```:

```groovy
def foo() {
    FormData formData = new FormData()
    ormData.addField('foo', null)
}
```

The reason for this is due to the ```add...``` methods having a ```(String, String)``` and ```(String, List)``` signature. Groovy is currently just picking one (unless the user expressly casts the type of ```null```), which picks the ```List``` signature, and then the attempt to add ```null``` to the **MultiValuedMap** results in the **NullPointerException**.

One remedy may be to perform a ```null```-check in both methdos and either ```return``` without action, or insert ```''``` (blank String).

A current work-around for Clients is to cast the value or to do like so:

```groovy
def foo(String bar) {
    HttpRequest httpRequest = new HttpRequest()
    if (bar) {
        httpRequest.addHeader('bar', bar)
    }
}
```

Or to use the ```set*``` methods, which execute different code in **MultiValuedMap** that does not call ```addAll``` and so does not throw a **NullPointerException** (note: Not available for **FormData**). However, these cases may not be as desirable, as the value may still be presented as the String ```'null'```, which may not be valid for the intent, which would be a blank (```''```) value.

Example stacktrace:

```text
Caused by: java.lang.NullPointerException: null
	at java.util.ArrayList.addAll(ArrayList.java:581)
	at com.budjb.httprequests.MultiValuedMap.add(MultiValuedMap.java:42)
	at com.budjb.httprequests.HttpRequest.addHeader(HttpRequest.java:279)
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,4 +1,1 @@
 language: java
-
-jdk:
-- oraclejdk8

--- a/gradle.properties
+++ b/gradle.properties
@@ -13,7 +13,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-version=2.0.4
+version=2.0.5
 group=com.budjb
 githubHttpsUrl=https://github.com/budjb/http-requests
 githubGitUrl=git@github.com:budjb/http-request.git

--- a/http-requests-core/src/main/java/com/budjb/httprequests/MultiValuedMap.java
+++ b/http-requests-core/src/main/java/com/budjb/httprequests/MultiValuedMap.java
@@ -19,49 +19,116 @@ package com.budjb.httprequests;
 import java.util.*;
 
 public class MultiValuedMap extends TreeMap<String, List<String>> {
+    /**
+     * Creates a new, empty multi-valued map.
+     */
     public MultiValuedMap() {
         super(String.CASE_INSENSITIVE_ORDER);
     }
 
+    /**
+     * Creates a new multi-valued map with the contents of the other, given map.
+     *
+     * @param other Other map to use as the basis for this map.
+     */
     public MultiValuedMap(Map<String, List<String>> other) {
         this();
         putAll(other);
     }
 
+    /**
+     * Adds the given value to the given key.
+     * <p>
+     * Note that this method does not protect against duplicate values.
+     *
+     * @param key   Key to add the given value to.
+     * @param value Value to add to the given key.
+     */
     public void add(String key, String value) {
         if (!containsKey(key)) {
             put(key, new ArrayList<>());
         }
-        get(key).add(value);
+
+        if (value != null) {
+            get(key).add(value);
+        }
     }
 
+    /**
+     * Adds the given values to the given key.
+     * <p>
+     * Note that this method does not protect against duplicate values.
+     *
+     * @param key    Key to add the given values to.
+     * @param values Values to add to the given key.
+     */
     public void add(String key, List<String> values) {
         if (!containsKey(key)) {
             put(key, new ArrayList<>());
         }
-        get(key).addAll(values);
+
+        if (values != null) {
+            get(key).addAll(values);
+        }
     }
 
+    /**
+     * Adds the contents of the given map to the existing contents of this map.
+     * <p>
+     * Note that this method does not protect against duplicate values.
+     *
+     * @param other Other map to merge into this map.
+     */
     public void add(Map<String, List<String>> other) {
-        other.forEach((k, v) -> {
-            if (containsKey(k)) {
-                get(k).addAll(v);
-            }
-            else {
-                put(k, v);
-            }
-        });
+        if (other != null) {
+            other.forEach((k, v) -> {
+                if (containsKey(k)) {
+                    get(k).addAll(v);
+                }
+                else {
+                    put(k, v);
+                }
+            });
+        }
     }
 
+    /**
+     * Sets and overwrites the value of the given key.
+     *
+     * @param key   Key to set the value of.
+     * @param value Value of the key.
+     */
     public void set(String key, String value) {
-        put(key, Arrays.asList(value));
+        if (value == null) {
+            put(key, new ArrayList<>());
+        }
+        else {
+            put(key, Arrays.asList(value));
+        }
     }
 
+    /**
+     * Sets and overwrites the values of the given key.
+     *
+     * @param key    Key to set the values of.
+     * @param values Values of the key.
+     */
     public void set(String key, List<String> values) {
-        put(key, values);
+        if (values == null) {
+            put(key, new ArrayList<>());
+        }
+        else {
+            put(key, values);
+        }
     }
 
+    /**
+     * Replaces the contents of this map with those of the given map.
+     *
+     * @param other The map to replace the contents of this map with.
+     */
     public void set(Map<String, List<String>> other) {
+        Objects.requireNonNull(other);
         clear();
         other.forEach(this::put);
     }
@@ -80,16 +147,17 @@ public class MultiValuedMap extends TreeMap<String, List<String>> {
         return map;
     }
 
+    /**
+     * Returns the flattened values of the given key, where each value is joined by a comma.
+     *
+     * @param key Key to retrieve the flattened values of.
+     * @return The flattened value.
+     */
     public String getFlat(String key) {
         if (!containsKey(key)) {
             return null;
         }
 
         return String.join(",", get(key));
-    }
-
-    @Override
-    public void clear() {
-        super.clear();
     }
 }

--- a/http-requests-core/src/test/groovy/com/budjb/httprequests/MultiValuedMapSpec.groovy
+++ b/http-requests-core/src/test/groovy/com/budjb/httprequests/MultiValuedMapSpec.groovy
@@ -1,0 +1,77 @@
+/*
+ * Copyright 2016-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.budjb.httprequests
+
+import spock.lang.Specification
+
+class MultiValuedMapSpec extends Specification {
+    def 'Setting a null value to a key that does not exist in the map results in an entry with no value'() {
+        setup:
+        MultiValuedMap map = new MultiValuedMap();
+
+        when:
+        map.set('foo', null)
+
+        then:
+        (Map) map == [
+            foo: []
+        ]
+    }
+
+    def 'Adding null values to a key that does not exist in the map results in an entry with no value'() {
+        setup:
+        MultiValuedMap map = new MultiValuedMap();
+
+        when:
+        map.add('foo', null)
+
+        then:
+        (Map) map == [
+            foo: []
+        ]
+    }
+
+    def 'Adding null values to a key with an existing value in the map has no effect'() {
+        setup:
+        MultiValuedMap map = new MultiValuedMap();
+        map.add('foo', 'bar')
+
+        when:
+        map.add('foo', null)
+
+        then:
+        (Map) map == [
+            foo: [
+                'bar'
+            ]
+        ]
+    }
+
+    def 'Setting a null value to a key with an existing value in the map results in an entry with no value'() {
+        setup:
+        MultiValuedMap map = new MultiValuedMap();
+        map.add('foo', 'bar')
+
+        when:
+        map.set('foo', null)
+
+        then:
+        (Map) map == [
+            foo: []
+        ]
+    }
+}

--- a/http-requests-documentation/src/docs/changelog.adoc
+++ b/http-requests-documentation/src/docs/changelog.adoc
@@ -2,6 +2,10 @@
 
 The changelog will be updated with release notes for each release of the library.
 
+2.0.5::
+
+* Add handling for `null` values in multi-valued maps (which affects request headers, query parameters, and form fields).
+
 2.0.4::
 
 * Update Gradle version and build dependencies.

--- a/http-requests-functional-test/src/main/groovy/com/budjb/httprequests/HttpIntegrationTestSuiteSpec.groovy
+++ b/http-requests-functional-test/src/main/groovy/com/budjb/httprequests/HttpIntegrationTestSuiteSpec.groovy
@@ -570,7 +570,6 @@ abstract class HttpIntegrationTestSuiteSpec extends AbstractIntegrationSpec {
         filter.closed
     }
 
-
     def 'When a URL does not exist and the response has no entity, a 404 status code should be captured with no exception'() {
         when:
         HttpResponse response = httpClientFactory.createHttpClient().get("${baseUrl}/test404")
@@ -578,6 +577,32 @@ abstract class HttpIntegrationTestSuiteSpec extends AbstractIntegrationSpec {
         then:
         response.status == 404
         response.getEntity(String) == null
+    }
+
+    def 'Query parameters that have been set but contain no values are still sent with the request URI'() {
+        setup:
+        HttpRequest request = new HttpRequest("${baseUrl}/testParams")
+        request.setQueryParameter('foo', null)
+
+        when:
+        Map response = httpClientFactory.createHttpClient().get(request).getEntity(Map)
+
+        then:
+        response.containsKey('foo')
+        !((List) response.get('foo'))[0]
+    }
+
+    def 'Headers that have been set but contain no value are still sent with the request'() {
+        setup:
+        HttpRequest request = new HttpRequest("${baseUrl}/testHeaders")
+        request.setHeader('x-foo', null)
+
+        when:
+        Map response = httpClientFactory.createHttpClient().get(request).getEntity(Map)
+
+        then:
+        response.containsKey('x-foo')
+        !((List) response.get('x-foo'))[0]
     }
 
     static class CloseableFilter implements HttpClientFilter, Closeable {

--- a/http-requests-functional-test/src/main/groovy/com/budjb/httprequests/application/TestApp.groovy
+++ b/http-requests-functional-test/src/main/groovy/com/budjb/httprequests/application/TestApp.groovy
@@ -16,6 +16,7 @@
 package com.budjb.httprequests.application
 
 import groovy.json.JsonBuilder
+import groovy.json.JsonOutput
 import org.springframework.boot.SpringApplication
 import org.springframework.boot.autoconfigure.SpringBootApplication
 import org.springframework.http.HttpHeaders
@@ -125,11 +126,13 @@ class TestApp {
 
     @RequestMapping(value = '/testParams', produces = 'application/json')
     String testParams(@RequestParam MultiValueMap<String, String> params) {
+        println JsonOutput.prettyPrint(JsonOutput.toJson(params))
         return new JsonBuilder(params).toString()
     }
 
     @RequestMapping(value = '/testHeaders', produces = 'application/json')
     String testHeaders(@RequestHeader MultiValueMap<String, String> headers) {
+        println JsonOutput.prettyPrint(JsonOutput.toJson(headers))
         return new JsonBuilder(headers).toString()
     }
 

--- a/http-requests-httpcomponents-client/src/main/java/com/budjb/httprequests/httpcomponents/client/HttpComponentsHttpClient.java
+++ b/http-requests-httpcomponents-client/src/main/java/com/budjb/httprequests/httpcomponents/client/HttpComponentsHttpClient.java
@@ -56,11 +56,25 @@ public class HttpComponentsHttpClient extends AbstractHttpClient {
 
         URIBuilder uriBuilder = new URIBuilder(request.getUri());
 
-        request.getQueryParameters().forEach((k, v) -> v.forEach(value -> uriBuilder.addParameter(k, value)));
+        request.getQueryParameters().forEach((k, v) -> {
+            if (v.size() == 0) {
+                uriBuilder.addParameter(k, null);
+            }
+            else {
+                v.forEach(value -> uriBuilder.addParameter(k, value));
+            }
+        });
 
         HttpUriRequest httpRequest = createHttpRequest(method, uriBuilder.build());
 
-        request.getHeaders().forEach((k, v) -> v.forEach(value -> httpRequest.addHeader(k, value)));
+        request.getHeaders().forEach((k, v) -> {
+            if (v.size() == 0) {
+                httpRequest.addHeader(k, null);
+            }
+            else {
+                v.forEach(value -> httpRequest.addHeader(k, value));
+            }
+        });
 
         if (httpEntity != null && httpRequest instanceof HttpEntityEnclosingRequest) {
             ContentType contentType = null;

--- a/http-requests-jersey1/src/main/java/com/budjb/httprequests/jersey1/JerseyHttpClient.java
+++ b/http-requests-jersey1/src/main/java/com/budjb/httprequests/jersey1/JerseyHttpClient.java
@@ -27,6 +27,7 @@ import com.sun.jersey.client.urlconnection.HTTPSProperties;
 import java.io.IOException;
 import java.io.OutputStream;
 import java.security.GeneralSecurityException;
+import java.util.List;
 
 /**
  * An implementation of {@link HttpClient} that uses the Jersey Client 1.x library.
@@ -138,8 +139,15 @@ public class JerseyHttpClient extends AbstractHttpClient {
      */
     private WebResource applyQueryParameters(WebResource resource, MultiValuedMap queryParameters) {
         for (String name : queryParameters.keySet()) {
-            for (String value : queryParameters.get(name)) {
-                resource = resource.queryParam(name, value);
+            List<String> values = queryParameters.get(name);
+
+            if (values.size() == 0) {
+                resource = resource.queryParam(name, "");
+            }
+            else {
+                for (String value : queryParameters.get(name)) {
+                    resource = resource.queryParam(name, value);
+                }
             }
         }
 
@@ -155,8 +163,15 @@ public class JerseyHttpClient extends AbstractHttpClient {
      */
     private WebResource.Builder applyHeaders(WebResource.Builder builder, MultiValuedMap headers) {
         for (String name : headers.keySet()) {
-            for (String value : headers.get(name)) {
-                builder = builder.header(name, value);
+            List<String> values = headers.get(name);
+
+            if (values.size() == 0) {
+                builder = builder.header(name, "");
+            }
+            else {
+                for (String value : headers.get(name)) {
+                    builder = builder.header(name, value);
+                }
             }
         }
 

--- a/http-requests-jersey2/src/main/java/com/budjb/httprequests/jersey2/JerseyHttpClient.java
+++ b/http-requests-jersey2/src/main/java/com/budjb/httprequests/jersey2/JerseyHttpClient.java
@@ -30,6 +30,7 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.security.GeneralSecurityException;
+import java.util.List;
 
 public class JerseyHttpClient extends AbstractHttpClient {
     /**
@@ -129,8 +130,15 @@ public class JerseyHttpClient extends AbstractHttpClient {
      */
     private WebTarget applyQueryParameters(WebTarget target, MultiValuedMap queryParameters) {
         for (String name : queryParameters.keySet()) {
-            for (String value : queryParameters.get(name)) {
-                target = target.queryParam(name, value);
+            List<String> values = queryParameters.get(name);
+
+            if (values.size() == 0) {
+                target = target.queryParam(name, "");
+            }
+            else {
+                for (String value : queryParameters.get(name)) {
+                    target = target.queryParam(name, value);
+                }
             }
         }
 
@@ -146,8 +154,15 @@ public class JerseyHttpClient extends AbstractHttpClient {
      */
     private Invocation.Builder applyHeaders(Invocation.Builder builder, MultiValuedMap headers) {
         for (String name : headers.keySet()) {
-            for (String value : headers.get(name)) {
-                builder = builder.header(name, value);
+            List<String> values = headers.get(name);
+
+            if (values.size() == 0) {
+                builder = builder.header(name, "");
+            }
+            else {
+                for (String value : headers.get(name)) {
+                    builder = builder.header(name, value);
+                }
             }
         }
 

